File: store-java/src/main/java/io/mb/store/service/StoreService.java
package io.mb.store.service;

import io.mb.store.model.Item;
import io.mb.store.repository.StoreRepository;
import io.mb.store.resource.dto.ItemCreateDto;
import io.mb.store.resource.dto.ItemDto;
import io.mb.store.resource.mapper.ItemCreateDtoMapper;
import io.mb.store.resource.mapper.ItemMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class StoreService {
    private final StoreRepository storeRepository;
    private final ItemMapper itemMapper;
    private final ItemCreateDtoMapper itemCreateDtoMapper;

    public List<ItemDto> findAll() {
        return storeRepository.findAll().stream()
                .map(itemMapper::toItemDto).toList();
    }

    public List<ItemDto> findByDealerId(final String dealerId) {
        return storeRepository.findByDealerId(dealerId).stream()
                .map(itemMapper::toItemDto).toList();
    }

    public Map<String, List<ItemDto>> finAllGroupedByDealerId() {
        return storeRepository.findAll().stream()
                .map(itemMapper::toItemDto)
                .collect(Collectors.groupingBy(ItemDto::dealerId));
    }

    @Transactional
    public ItemDto save(final ItemCreateDto itemCreateDto) {
        Item savedItem = storeRepository.save(itemCreateDtoMapper.mapToItem(itemCreateDto));
        return itemMapper.toItemDto(savedItem);
    }
}

File: store-kotlin/src/main/kotlin/io/mb/store/service/StoreService.kt
package io.mb.store.service

import io.mb.store.model.Item
import io.mb.store.repository.StoreRepository
import io.mb.store.resource.dto.ItemCreateDto
import io.mb.store.resource.dto.ItemDto
import io.mb.store.resource.mapper.toItem
import io.mb.store.resource.mapper.toItemDto
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class StoreService(
    private val storeRepository: StoreRepository,
) {

    fun findAll(): List<ItemDto> =
        storeRepository.findAll().map(Item::toItemDto)

    fun findByDealerId(dealerId: String): List<ItemDto> =
        storeRepository.findByDealerId(dealerId)
            .map(Item::toItemDto)

    /**
     * TODO (DONE) Implement the method below
     *
     * This method should return all items grouped by dealer id
     */
    fun findAllGroupedByDealerId(): Map<String, List<ItemDto>> =
        storeRepository.findAll().asSequence()
            .map(Item::toItemDto)
            .groupBy { it.dealerId }

    @Transactional
    fun save(itemCreateDto: ItemCreateDto): ItemDto =
        storeRepository.save(itemCreateDto.toItem()).toItemDto()
}